### PR TITLE
virt-handler, VMI iface status: Fix detached interfaces status reporting

### PR DESCRIPTION
**Pull request by ormergi**
**What this PR does / why we need it**:
Currently an unplugged iface reporting is flaky due to a race between virt-handler and virt-controller, where virt-handler set the VMI iface status according to the domain which dosent include detached interfaces.

On the other hande virt-controller see the subject external interface stil exist (according to the CNI annotation) and add it to the VMI status.

The end result is unpluged interfaces status is flaky, where it keeps get deleted and added.

To fix this, keep detached interfaces status whos  external iface exist in pod (have info-source mutlus-status) and also exist in the spec.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

**Special notes for your reviewer**:

**Release note**:

```release-note
NONE
```

### CODE REVIEW

File: pkg/network/setup/netstat.go [hunk 1]
```diff
@@ -138,6 +138,18 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
 		}
 	}
 
+	indexedIfaceStatus := netvmispec.IndexInterfacesFromStatus(interfacesStatus, func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool { return true })
```
Comment (EdDev) [inline pkg/network/setup/netstat.go:141]: I think you just need to adjust the previous iteration (on line 134).

Instead of iterating over the `interfacesStatus`, there is a need to iterate over the `multusStatusNetworksByName` and perform a lookup for the interface in the status.
If it is there, add the source (as now).
If it is not there, add a new entry as you do in this new block.

Having two iterations is confusing and redundant IMO.

**Comment by ormergi** [comment]
[Changes](https://github.com/kubevirt/kubevirt/compare/e15cceb92b3178254469ce6ed58b14b6a70431f2..3d84fa76fb9d4ebe5dc5e0db91317e81f1c84646): fix review https://github.com/kubevirt/kubevirt/pull/10063#discussion_r1257489007

File: pkg/network/setup/netstat_test.go [hunk 2]
```diff
@@ -741,6 +778,30 @@ func (t *testSetup) addNetworkInterface(vmiIface v1.Interface, vmiNetwork v1.Net
 	return nil
 }
 
+// addDetachedInterface adds interface that does not exist in domain spec (detached iface),
+// but has the wiring set in virt-launcher pod (e.g.: in case of bridge binding, tap device and linux bridge)
+// This consist of 2 entities and pod volatile cache:
+// - vmi spec interface
+// - vmi spec network
+// - virt-launcher/pod filesystem cache
```
Comment (AlonaKaplan) [inline pkg/network/setup/netstat_test.go:786]: same.

File: pkg/network/setup/netstat.go [hunk 3]
```diff
@@ -131,10 +131,19 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
-	for ifaceIndex, ifaceStatus := range interfacesStatus {
-		if _, exists := multusStatusNetworksByName[ifaceStatus.Name]; exists {
-			interfacesStatus[ifaceIndex].InfoSource = netvmispec.AddInfoSource(
-				ifaceStatus.InfoSource, netvmispec.InfoSourceMultusStatus)
+	indexedIfaceStatus := netvmispec.IndexInterfacesFromStatus(interfacesStatus,
+		func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool { return true })
+	for multusIfaceName := range multusStatusNetworksByName {
+		_, existInSpec := vmiInterfacesSpecByName[multusIfaceName]
+		_, existInNewStatus := indexedIfaceStatus[multusIfaceName]
+		if existInSpec && !existInNewStatus {
```
[the remark this answers is not in the document]
Comment (EdDev) [reply pkg/network/setup/netstat.go:139]: I think we cannot depend on the status, we have been there before and it was found problematic.
It will be better to limit the dependency as much as possible, fetching information from the previous status with great care.

It will be enough to add an existing dependency in the virt-controller, to "leak" entries.

Going through the spec is a small attempt to avoid such a potential "leak" and reduce the dependency.

File: pkg/network/setup/netstat.go [hunk 4]
```diff
@@ -131,10 +131,19 @@ func (c *NetStat) UpdateStatus(vmi *v1.VirtualMachineInstance, domain *api.Domai
...
+			interfacesStatus = append(interfacesStatus, v1.VirtualMachineInstanceNetworkInterface{
+				Name:       multusIfaceName,
+				InfoSource: netvmispec.InfoSourceMultusStatus,
+			})
+		} else if existInNewStatus {
+			iface := netvmispec.LookupInterfaceStatusByName(interfacesStatus, multusIfaceName)
```
Comment (EdDev) [inline pkg/network/setup/netstat.go:145]: This is the only lookup needed, the `indexedIfaceStatus` is not needed, it is a duplication AFAIU.

File: pkg/network/setup/netstat.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go]: You right, its not a must and works w/o it.
I thought it should be added because of this [test](https://github.com/kubevirt/kubevirt/blob/main/pkg/network/setup/netstat_test.go#L205), if fails if we don't check the iface exists in spec.

But I guess it's fine and the test should be changed because now we do expect to see such interfaces in the status.

Removed the checking against the spec and changed [unit-test](https://github.com/kubevirt/kubevirt/blob/main/pkg/network/setup/netstat_test.go#L205).

[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go]: I missed this comment https://github.com/kubevirt/kubevirt/pull/10063#discussion_r1258310610

File: pkg/network/setup/netstat.go (hunk 4 above)
Comment (ormergi) [reply pkg/network/setup/netstat.go]: This lookup is performed to mutate the actual interfaces that will be set on the VMI object.
Mutating `indexedIfaceStatus` iteration objected won't  because it's not the same underlying object

File: pkg/network/setup/netstat.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/setup/netstat.go]: As Edy mentioned, initially it was an attempt to not rely on the previous status entirely.

File: pkg/network/setup/netstat.go (hunk 4 above)
Comment (EdDev) [reply pkg/network/setup/netstat.go:145]: I think the comment was about not needing the indexing and just doing this lookup. Checking if it exists or not, if it exists it is updated, if it does not it is added.

Comment (EdDev) [reply pkg/network/setup/netstat.go:145]: I am writing the whole suggestion here to be clear.
Note, this whole block needs to come in a function like all the other source collections above.

```go
func ifacesStatusFromMultus(interfacesStatus []v1.VirtualMachineInstanceNetworkInterface, multusStatusNetworksByName map[string]blabla, vmIfacesSpecByName map[string]blabla) []v1.VirtualMachineInstanceNetworkInterface {
	for ifaceName := range multusStatusNetworksByName {
	        ifaceStatus := netvmispec.LookupInterfaceStatusByName(interfacesStatus, ifaceName)
                _, existInSpec := vmIfacesSpecByName[ifaceName]
                if existInSpec && ifaceStatus == nil {
			interfacesStatus = append(interfacesStatus, v1.VirtualMachineInstanceNetworkInterface{
				Name:       ifaceName,
				InfoSource: netvmispec.InfoSourceMultusStatus,
			})
		} else if ifaceStatus != nil {
			ifaceStatus.InfoSource = netvmispec.AddInfoSource(ifaceStatus.InfoSource, netvmispec.InfoSourceMultusStatus)
		}
	}
        return interfacesStatus
}
```

File: pkg/network/setup/netstat_test.go [hunk 5]
```diff
@@ -302,6 +302,40 @@ var _ = Describe("netstat", func() {
...
+			Expect(setup.NetStat.UpdateStatus(setup.Vmi, setup.Domain)).To(Succeed())
+
+			infoSourceDomainGAMultus := netvmispec.NewInfoSource(
+				netvmispec.InfoSourceDomain, netvmispec.InfoSourceGuestAgent, netvmispec.InfoSourceMultusStatus)
+			Expect(setup.Vmi.Status.Interfaces).To(Equal([]v1.VirtualMachineInstanceNetworkInterface{
+				newVMIStatusIface(primaryNetworkName, []string{primaryGaIPv4, primaryGaIPv6}, primaryMAC, primaryIfaceName, infoSourceDomainGAMultus, netsetup.DefaultInterfaceQueueCount),
+				newVMIStatusIface(secondaryNetworkName, nil, "", "", netvmispec.InfoSourceMultusStatus, 0),
+			}), "primary and secondary ifaces should exist in status, where secondary iface have multus-status only")
+
+			Expect(setup.NetStat.PodInterfaceVolatileDataIsCached(setup.Vmi, primaryNetworkName)).To(BeTrue())
```
Comment (EdDev) [inline pkg/network/setup/netstat_test.go:337]: And what do we expect from the secondary?

File: pkg/network/setup/netstat_test.go [hunk 6]
```diff
@@ -302,6 +302,40 @@ var _ = Describe("netstat", func() {
 			Expect(setup.NetStat.PodInterfaceVolatileDataIsCached(setup.Vmi, primaryNetworkName)).To(BeFalse())
 			Expect(setup.NetStat.PodInterfaceVolatileDataIsCached(setup.Vmi, secondaryNetworkName)).To(BeFalse())
 		})
+
+		It("run status and expect 1 attached iface & 1 detached iface to be reported based on multus status and guest-agent data", func() {
```
Comment (EdDev) [inline pkg/network/setup/netstat_test.go:306]: I think this is not covering all the paths of the new code.

Looking at the logic, we need also: Appears in the original status as multus, but it is no longer in the spec.

**Comment by ormergi** [comment]
@AlonaKaplan @EdDev some more changes were necessary around filtering interfaces for hotplug (also for SR-IOV interfaces), please take another look.

File: pkg/network/vmispec/interface_test.go [hunk 7]
```diff
@@ -94,15 +94,6 @@ var _ = Describe("VMI network spec", func() {
 
 	const iface1, iface2, iface3, iface4, iface5 = "iface1", "iface2", "iface3", "iface4", "iface5"
 
-	DescribeTable("return VMI spec interface names, given",
```
Comment (AlonaKaplan) [inline pkg/network/vmispec/interface_test.go:97]: Why was this test removed?

File: pkg/virt-launcher/virtwrap/nichotplug.go [hunk 8]
```diff
@@ -113,9 +113,7 @@ func (vim *virtIOInterfaceManager) hotUnplugVirtioInterface(vmi *v1.VirtualMachi
 }
 
 func interfacesToHotUnplug(vmiSpecInterfaces []v1.Interface, domainSpecInterfaces []api.Interface) []api.Interface {
-	ifaces2remove := netvmispec.FilterInterfacesSpec(vmiSpecInterfaces, func(i v1.Interface) bool {
-		return i.State == v1.InterfaceStateAbsent
-	})
+	ifaces2remove := netvmispec.FilterIfacesForHotUnplug(vmiSpecInterfaces)
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/nichotplug.go:116]: Only "bridge" interfaces can be "absent", so I find this change redundant.

File: pkg/virt-launcher/virtwrap/nichotplug.go [hunk 9]
```diff
@@ -142,16 +140,19 @@ func lookupDomainInterfaceByName(domainIfaces []api.Interface, networkName strin
...
 
 func networksToHotplugWhoseInterfacesAreNotInTheDomain(vmi *v1.VirtualMachineInstance, indexedDomainIfaces map[string]api.Interface) []v1.Network {
+	ifacesToHotplug := netvmispec.FilterIfacesForHotplug(vmi.Spec.Domain.Devices.Interfaces)
+	pluggableIfaces := netvmispec.IndexInterfaceSpecByName(ifacesToHotplug)
+
 	var networksToHotplug []v1.Network
 	interfacesToHoplug := netvmispec.IndexInterfacesFromStatus(
 		vmi.Status.Interfaces,
 		func(ifaceStatus v1.VirtualMachineInstanceNetworkInterface) bool {
-			_, exists := indexedDomainIfaces[ifaceStatus.Name]
```
Comment (AlonaKaplan) [inline pkg/virt-launcher/virtwrap/nichotplug.go:149]: I think the same fix as was done in virt-handler should be applied here. Checking the info source.

File: pkg/network/vmispec/interface_test.go (hunk 7 above)
Comment (ormergi) [reply pkg/network/vmispec/interface_test.go:97]: It was a mistake having it in the first place because there is no logic to test, all InterfaceName() does is indexing the interface since by name.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:149]: In virt-launcher context, the most accurate indication of whether an interface is plugged is the domain object, the status may not reflect the actual state at that expect point of time.

I am not sure why its necessary to check the status at all, its possible to attach an interface whether the external interface exist or not (add by the CNI), isn't it?

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 8 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:116]: It may seem premature but I think it is worth having it to prevent from other interface types being subject to unplug, it may also serve the futuristic SR-IOV hotplug/unplug support.
WDTY?

Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:116]: Absent SRIOV interface will have to be unplugged here as well. Let's keep the code generic.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:149]: I didn't mean you should remove the domain check. I meant using the info-source instead of checking the binding.

> I am not sure why its necessary to check the status at all, its possible to attach an interface whether the external interface exist or not (add by the CNI), isn't it?
No,  the hotplug is serial. The interface can be attached to the guest only after it was attached to the pod and the tap device was created.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 8 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:116]: SR-IOV interfaces cannot be handled at this place because its backed by hostDevice and not a regular interface in the domain spec.

File: pkg/network/setup/netstat.go (hunk 4 above)
Comment (ormergi) [reply pkg/network/setup/netstat.go]: Thanks for the example, done

File: pkg/network/setup/netstat_test.go (hunk 6 above)
Comment (ormergi) [reply pkg/network/setup/netstat_test.go:306]: Added test for the suggested scenario

File: pkg/network/vmispec/interface_test.go (hunk 7 above)
Comment (ormergi) [reply pkg/network/vmispec/interface_test.go:97]: This function is removed altogether in favor of using a helper that already exists.

File: pkg/virt-handler/vm.go [hunk 10]
```diff
@@ -3049,8 +3049,14 @@ func (d *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
 
 func (d *VirtualMachineController) hotplugSriovInterfaces(vmi *v1.VirtualMachineInstance) error {
 	sriovSpecInterfaces := netvmispec.FilterSRIOVInterfaces(vmi.Spec.Domain.Devices.Interfaces)
-	sriovStatusInterfaces := netvmispec.FilterStatusInterfacesByNames(vmi.Status.Interfaces, netvmispec.InterfacesNames(sriovSpecInterfaces))
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-handler/vm.go:3052]: Done

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:149]: As we concluded on an offline discussion, checking the binding is crucial for the scenario SR-IOV interfaces exist, because the checking if interface exist in domain will fail because SR-IOV interfaces are backed by host-devices.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 8 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:116]: I think you meant for the filtering in virt-handler, if so I removed that commit as we concluded offline, in favor of keeping the implementation generic at virt-handler level.

File: pkg/virt-handler/vm.go [hunk 11]
```diff
@@ -3049,8 +3049,14 @@ func (d *VirtualMachineController) vmUpdateHelperDefault(origVMI *v1.VirtualMach
...
-	if len(sriovSpecInterfaces) == len(sriovStatusInterfaces) {
+
+	sriovSpecIfacesNames := netvmispec.InterfacesNames(sriovSpecInterfaces)
+	attachedSriovStatusIfaces := netvmispec.IndexInterfacesFromStatus(vmi.Status.Interfaces, func(iface v1.VirtualMachineInstanceNetworkInterface) bool {
+		_, exist := sriovSpecIfacesNames[iface.Name]
+		return netvmispec.ContainsInfoSource(iface.InfoSource, netvmispec.InfoSourceDomain) && exist
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/virt-handler/vm.go]: Done

File: pkg/network/vmispec/interface.go [hunk 12]
```diff
@@ -129,13 +129,14 @@ func LookupInterfaceByName(ifaces []v1.Interface, name string) *v1.Interface {
 	return nil
 }
 
-// InterfacesNames returns slice with the names of the given interfaces.
-func InterfacesNames(interfaces []v1.Interface) []string {
-	var ifaceNames []string
+// InterfacesNames returns map with the names of the given interfaces.
+func InterfacesNames(interfaces []v1.Interface) map[string]struct{} {
```
[the remark this answers is not in the document]
Comment (ormergi) [reply pkg/network/vmispec/interface.go]: Done

File: pkg/network/vmispec/hotplug.go [hunk 13]
```diff
@@ -69,3 +69,15 @@ func NetworksToHotplugWhosePodIfacesAreReady(vmi *v1.VirtualMachineInstance) []v
...
+	return FilterInterfacesSpec(ifaces, func(iface v1.Interface) bool {
+		return iface.State != v1.InterfaceStateAbsent && iface.SRIOV == nil
+	})
+}
+
+func FilterIfacesForHotUnplug(ifaces []v1.Interface) []v1.Interface {
+	return FilterInterfacesSpec(ifaces, func(iface v1.Interface) bool {
+		return iface.State == v1.InterfaceStateAbsent && iface.SRIOV == nil
+	})
+}
```
Comment (EdDev) [inline pkg/network/vmispec/hotplug.go:83]: Placing these helpers here is confusing. We will/do support hot-plug/unplug for SR-IOV, but one that reads this may assume we do not.

I think we can use the generic filter function where it is needed, because at that location SR-IOV is not treated. It will also remove the tests and unneeded boilerplate which comes with these.

File: pkg/virt-launcher/virtwrap/nichotplug.go [hunk 14]
```diff
@@ -142,16 +140,19 @@ func lookupDomainInterfaceByName(domainIfaces []api.Interface, networkName strin
 }
 
 func networksToHotplugWhoseInterfacesAreNotInTheDomain(vmi *v1.VirtualMachineInstance, indexedDomainIfaces map[string]api.Interface) []v1.Network {
+	ifacesToHotplug := netvmispec.FilterIfacesForHotplug(vmi.Spec.Domain.Devices.Interfaces)
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/nichotplug.go:143]: I have hard time to understand the need for the changes in this function.
I would like to see the logical change needed here, but it is very hard to follow it with all the refactoring done.

I think, that we do not need all the indexing and filtering added, just a single additional condition to the predicate of `IntexInterfacesFromStatus`.
If you feel it requires refactoring, please do it in a follow up PR (doing this in this PR will delay its acceptance).

File: pkg/virt-launcher/virtwrap/nichotplug_test.go [hunk 15]
```diff
@@ -116,6 +120,25 @@ var _ = Describe("nic hotplug on virt-launcher", func() {
 			map[string]api.Interface{},
 			[]v1.Network{generateNetwork(networkName, nadName)},
 		),
+		Entry("given vmi with SR-IOV interfaces, should return no network",
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/nichotplug_test.go:123]: > given vmi with SR-IOV interfaces, should return no network

Please keep the same format as the other test descriptions.
E.g.:
> vmi with 1 SR-IOV network (when the pod interface *is* ready) and no interfaces in the domain

**Review by EdDev — CHANGES_REQUESTED** [review]

File: pkg/network/vmispec/hotplug.go [hunk 16]
```diff
@@ -69,3 +69,15 @@ func NetworksToHotplugWhosePodIfacesAreReady(vmi *v1.VirtualMachineInstance) []v
 
 	return networksToHotplug
 }
+
```
Comment (AlonaKaplan) [inline pkg/network/vmispec/hotplug.go:72]: Please fix the commit message.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 14 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:143]: I totally agree.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/nichotplug.go:149]: For tracking history, please mention in the commit message that the change to the virt-handler status calculation didn't cause the issue this commit fixes. It was there before as well.

File: pkg/network/setup/netstat_test.go (hunk 5 above)
Comment (ormergi) [reply pkg/network/setup/netstat_test.go:337]: I think we should not care for the secondary because it describes a scenario it is detached.

File: pkg/network/vmispec/hotplug.go (hunk 16 above)
Comment (ormergi) [reply pkg/network/vmispec/hotplug.go:72]: Whats wrong with it?

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:149]: Sorry I don't follow, what you mean by " the change to the virt-handler status calculation"?

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 14 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:143]: Adding another condition to the predicate is making the code worse instead of the suggested changes.

I will post a PR for refactoring this..

File: pkg/network/vmispec/hotplug.go (hunk 13 above)
Comment (ormergi) [reply pkg/network/vmispec/hotplug.go:83]: These are mere wrappers for the logic needed to get the interfaces to un/plug, with no boilerplate whatsoever.

Note that having these functions enable testing the filtering logic of interfaces for un/plug, testing under vm_tests.go is not always possible and takes way more time.

File: pkg/virt-launcher/virtwrap/nichotplug_test.go (hunk 15 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug_test.go:123]: The current format is odd because it doesn't express the expected result: 
`<fn name> given ...`

I will change it to what you suggest.

File: pkg/network/vmispec/hotplug.go (hunk 13 above)
Comment (EdDev) [reply pkg/network/vmispec/hotplug.go:83]: > with no boilerplate whatsoever.

You can count the number of code lines that got removed once you just do the minimum as suggested. That is the boilerplate (e.g. wrapper function, tests for the wrapper function).

> Note that having these functions enable testing the filtering logic of interfaces for un/plug, testing under vm_tests.go is not always possible and takes way more time.

What interfaces to plug and unplug is not limited to the logic in these functions.

In the end it is all about balance. I think these helpers are not worth it.

File: pkg/network/setup/netstat_test.go (hunk 5 above)
Comment (EdDev) [reply pkg/network/setup/netstat_test.go:337]: You have it as input in this test and I think there is a need to express what is expected of it on way or the other.

File: pkg/network/vmispec/hotplug.go (hunk 16 above)
Comment (EdDev) [reply pkg/network/vmispec/hotplug.go:72]: > Filter interfaces with bridge binding only for hot{unp}plug operations as its the
only supposed binding at the moment.

This ^^

Comment (ormergi) [reply pkg/network/vmispec/hotplug.go]: Done

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (AlonaKaplan) [reply pkg/virt-launcher/virtwrap/nichotplug.go]: The first commit in this PR - `virt-handler, VMI iface status: Keep detached secondary ifaces status`.

I will clarify my comment - commit `virt-launcher, iface hotplug: Filter out unsupported ifaces` could be a standalone PR. The issue it fixes ("SyncVMI routine fails, causing virt-handler to keep re-enqueue the request.") wasn't caused by the first commit. I existed before.
Also the commit message says- "Migrated VMs with SR-IOV interfaces are also affected, their SR-IOV interfaces wont plug back after migration.". I don't think it is the case, I don't understand how the syncVMI can cause it.

File: pkg/virt-launcher/virtwrap/nichotplug.go [hunk 17]
```diff
@@ -113,8 +113,8 @@ func (vim *virtIOInterfaceManager) hotUnplugVirtioInterface(vmi *v1.VirtualMachi
 }
 
 func interfacesToHotUnplug(vmiSpecInterfaces []v1.Interface, domainSpecInterfaces []api.Interface) []api.Interface {
-	ifaces2remove := netvmispec.FilterInterfacesSpec(vmiSpecInterfaces, func(i v1.Interface) bool {
-		return i.State == v1.InterfaceStateAbsent
+	ifaces2remove := netvmispec.FilterInterfacesSpec(vmiSpecInterfaces, func(iface v1.Interface) bool {
+		return iface.State == v1.InterfaceStateAbsent && iface.SRIOV == nil
```
Comment (EdDev) [inline pkg/virt-launcher/virtwrap/nichotplug.go:117]: I think you need to cover this condition in the unit tests (under `DescribeTable("domain interfaces to hot-unplug",`).

File: pkg/network/setup/netstat_test.go [hunk 18]
```diff
@@ -302,6 +302,70 @@ var _ = Describe("netstat", func() {
...
+
+			Expect(setup.NetStat.UpdateStatus(setup.Vmi, setup.Domain)).To(Succeed())
+
+			infoSourceDomainGAMultus := netvmispec.NewInfoSource(
+				netvmispec.InfoSourceDomain, netvmispec.InfoSourceGuestAgent, netvmispec.InfoSourceMultusStatus)
+			Expect(setup.Vmi.Status.Interfaces).To(Equal([]v1.VirtualMachineInstanceNetworkInterface{
+				newVMIStatusIface(primaryNetworkName, []string{primaryGaIPv4, primaryGaIPv6}, primaryMAC, primaryIfaceName, infoSourceDomainGAMultus, netsetup.DefaultInterfaceQueueCount),
+			}), "only primary should exist in status since secondary iface not exist in spec")
+
+			Expect(setup.NetStat.PodInterfaceVolatileDataIsCached(setup.Vmi, primaryNetworkName)).To(BeTrue())
```
Comment (EdDev) [inline pkg/network/setup/netstat_test.go:367]: same

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 9 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go]: > I will clarify my comment - commit virt-launcher, iface hotplug: Filter out unsupported ifaces could be a standalone PR. The issue it fixes ("SyncVMI routine fails, causing virt-handler to keep re-enqueue the request.") wasn't caused by the first commit. I existed before.

It cannot be standalone because having this change requires filtering host-devices (i.e.: SR-IOV interfaces) at virt-launcher level on hot-un/plug flow, otherwise, virt-handler gets into re-enqueues storm;
because SyncVMI routine tries  to hot-un/plug SR-IOV interfaces and fails to fetch in Libevirt XML spec.

> Also the commit message says- "Migrated VMs with SR-IOV interfaces are also affected, their SR-IOV interfaces wont plug back after migration.". I don't think it is the case, I don't understand how the syncVMI can cause it.

Migrated VMs also affected due to what is described above, the regular hot-un/plug flow will try to plug SR-IOV interfaces whether they are plugged or not.

File: pkg/network/setup/netstat_test.go (hunk 5 above)
Comment (AlonaKaplan) [reply pkg/network/setup/netstat_test.go:337]: I agree will Or, the Volatile cache is not relevant in case of detached interface. It may contain the interface or may not we don't care.

Comment (AlonaKaplan) [reply pkg/network/setup/netstat_test.go:337]: But since I looked at his code again, it may be nice to add `setup.AddMultusOnlyInterface`:)

Comment (EdDev) [reply pkg/network/setup/netstat_test.go:337]: I think the test needs to express the expectation explicitly for a specific scenario.
I think it is important to express what the volatile cache will be when an interface is detached.

E.g. it is not touched, it is left as before.

File: pkg/virt-launcher/virtwrap/nichotplug.go (hunk 17 above)
Comment (ormergi) [reply pkg/virt-launcher/virtwrap/nichotplug.go:117]: Done

File: pkg/network/vmispec/hotplug_test.go [hunk 19]
```diff
@@ -105,6 +105,7 @@ var _ = Describe("utilitary funcs to identify attachments to hotplug", func() {
 			dummyVMIWithAttachmentAlreadyAvailableOnDomain(networkName, nadName, guestIfaceName),
 		),
 	)
+
```
Comment (EdDev) [inline pkg/network/vmispec/hotplug_test.go:108]: leftover

**Review by EdDev — APPROVED** [review]

File: pkg/network/setup/netstat_test.go (hunk 5 above)
Comment (ormergi) [reply pkg/network/setup/netstat_test.go:337]: Test moved outside the Conetxt() that tests volatile cache

File: pkg/network/vmispec/hotplug_test.go (hunk 19 above)
Comment (ormergi) [reply pkg/network/vmispec/hotplug_test.go:108]: Thanks, fixed.

**Review by EdDev — APPROVED** [review]
